_posts/2018-02-14-how-android-builds-work-in-bazel.md

```diff
@@ -0,0 +1,331 @@
+2.  In the __analysis__ phase, Bazel builds a graph of actions needed to build
+    the specified targets.
+3.  In the __execution__ phase, Bazel runs those actions.
+
+Each Bazel target is defined by a __rule__, which specifies inputs, outputs, and
+how to get from one to the other. Rules can specify things like creating an
+executable binary or defining a library. In Bazel’s code, individual rules are
+represented by instances of implementations of
+[`RuleConfiguredTargetFactory`](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/RuleConfiguredTargetFactory.java).
+Users can also extend Bazel and create new rules with
+[Skylark](https://docs.bazel.build/versions/master/skylark/concepts.html).
+
+Rules create, in turn, any number of __action__s. Each action takes any number
+of __artifact__s as inputs and produces one or more artifacts as outputs. These
+artifacts represent files that may not yet be available. They can either be
+_source artifacts_, such as source code checked in to the repository, or
+_generated artifacts_, such as output of other actions. For example, an action
+to compile a piece of code might take in source artifacts representing the code
+to be compiled and generated artifacts representing compiled dependencies, even
+though those dependencies have not yet been compiled, and output a generated
+artifact representing the compiled result. Additionally, rules may expose any
+number of __provider__ objects. These providers are the API rules provide to
+other rules. They provide read-only information about internal state.
+
+During analysis, Bazel runs the rules for each target being built and their
+transitive dependencies. Each rule generates and records all the actions it
+depends on. Bazel won’t necessarily run all of those actions; if an action
+doesn’t end up being required, Bazel will just ignore it. __Skyframe__ is used
+to evaluate and cache the results of rules.
+
+Information from the rules, including artifacts representing the future output
+of actions, are made available to other rules through the rules’ providers. Each
+rule has access to its direct dependencies’ providers. Because most information
+passed between rules is actually transitive, providers make use of the
+[`NestedSet`](https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/collect/nestedset/NestedSet.java)
+class, a DAG-like data structure (it’s not actually a set!) made up of items and
+pointers to other (nested) `NestedSet` objects. `NestedSet`s are specially
+optimized to work efficiently for analysis. For part of a provider that
+represents some transitive state, for example, a trivial implementation might be
+to build a new list that contains the items for the current rule and each
+transitive dependency (for a chain of n transitive dependencies, that means we’d
+add $$n + (n - 1) + … 1 = O(n^2)$$ items to some list), but building a nested
```

Comment:
I don't think we do any conversion of math syntax, remove the `$$`?

Reply:
removed